### PR TITLE
Aquifet parser implementation

### PR DESCRIPTION
**Pull request by WesselDeZeeuw**
Hi everyone!

First of all a small introduction is needed. My name is Wessel and since september I am working at TNO together with Rohith, Frank and Kelvin on the OPM project. My first task is to implement the Fetkovich Aquifer. 

For this I created the required parsing and test files for this. This is my first time doing real software developement, so please give me all the feedback and support you can think of! This merge is for dealing with the AQUFETP keyword in the input file. 
Thanks for the feedback, hope to hear soon from you guys!

Wessel

### CODE REVIEW

**Comment by atgeirr** [comment]
Welcome to the OPM community, Wessel!

About the PR itself: @joakim-hove this is your domain of expertise, so I assume you will do a review.

**Comment by atgeirr** [comment]
I will add one comment though: you may want to squash the commits together to one or a few, at least the revert commits should be combined with the preceding ones (and disappear).

**Comment by joakim-hove** [comment]
jenkins build this please

File: src/opm/parser/eclipse/EclipseState/Aquifetp.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <opm/parser/eclipse/EclipseState/Aquifetp.hpp>
+
+
+namespace Opm{
+	
+  Aquifetp::Aquifetp(const EclipseState& eclState, const Deck& deck)
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Aquifetp.cpp]: Does not seem like the `eclState` parameter is used at all - I guess the constructor could have signature:
```
Aquifetp::Aquifetp(const Deck& deck);
```

File: src/opm/parser/eclipse/EclipseState/Aquifetp.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      if (!deck.hasKeyword("AQUFETP"))
+        return;
+
+      const auto& aquifetpKeyword = deck.getKeyword("AQUFETP");
+
+      for (auto& aquifetpRecord : aquifetpKeyword){
+
+          Aquifetp::AQUFETP_data data;
+
+          data.aquiferID = aquifetpRecord.getItem("AQUIFER_ID").template get<int>(0);
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Aquifetp.cpp]: Quite few of the items in the `AQUFETP` keyword have a defined default; the call:
```
 aquifetpRecord.getItem("P0").getSIDouble(0);
```
will blow up with an exception if the`P0` item has not been given a value in the input deck. In cases like this you must protect like this:
```
const auto& item = record.getItem("P0"):
if (item.hasValue(0))
    data.p0 = item.getSIDouble(0);
```

**Comment by joakim-hove** [comment]
Hello Wessel;

welcome to OPM. This will PR will be merged, but I have some comments which I woould like to address first:

**Member of EclipseState:** The way you have implemented it the `Aquifet` class is created in solitude based on an existing `EclipseState` and `Deck` instance. We try to contain the `Deck` datastructure within the opm-common module, the downstream modules like opm-grid and opm-simulators should only access the more processed `EclipseState` object. To achieve that the `Aquifet` instance must be owned by the `EclipseState`. 

**Testing:** You have added a test - that is very good; good if you also add brief testing of the empty `Aquifet` container - i.e. check that the `Aquifet` container is meaningfull also for a deck which does not contain the `AQUIFET` keyword at all.

**Default handling:** You must protect agains accessing unset values; as I have commented elseweher. If the pressure is defaulted you can not get hold of it before the simulator is ready with an intial reservoir pressure - that is *extremely ugly* - but not much you can do about it. You will have add an additional `bool` flag and do something like:

```
const auto& p0_item = record.getItem("P0");
if (p0_item.hasValue(0)) {
      data.p0 = p0_item.getSIDouble( 0 );
      data.valid_p0 = true;
} else
      data.valid_p0 = false;
```

and then calling scope needs to check `data.valid_p0` before accessing the `p0` value.

**Minor stylistic stuff:**
You have a very deep indentation (8 spaces?) and some braces in weird locations (spaces / tabs issue?). Please try to polish up your indentation to more closely resemble the rest of the code - I will not drag you through town on these issues.

File: src/opm/parser/eclipse/EclipseState/Aquifetp.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+          data.d0 = aquifetpRecord.getItem("DAT_DEPTH").getSIDouble(0);
+          data.C_t = aquifetpRecord.getItem("C_T").getSIDouble(0);
+          data.V0 = aquifetpRecord.getItem("V0").getSIDouble(0);
+          data.J = aquifetpRecord.getItem("PI").getSIDouble(0);
+
+          m_aqufetp.push_back(std::move(data));
+      	}
+  	}
+
+  	const std::vector<Aquifetp::AQUFETP_data>& Aquifetp::getAquifers() const
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Aquifetp.cpp]: Instead of this method I would prefer if you exposed `begin()` and `end()` methods were forwarded to the underlying `std::vector`.

**Comment by joakim-hove** [comment]
OK - no further comments for now; feel free to ask if you have questions.

File: CMakeLists_files.cmake [hunk 4]
```diff
@@ -429,6 +431,7 @@ if(ENABLE_ECL_INPUT)
        opm/parser/eclipse/EclipseState/EclipseConfig.hpp
        opm/parser/eclipse/EclipseState/Aquancon.hpp
        opm/parser/eclipse/EclipseState/AquiferCT.hpp
+       opm/parser/eclipse/EclipseState/Aquiferfetp.hpp
```
Comment (joakim-hove) [inline CMakeLists_files.cmake]: This is wrong - the file is just called `Aquifetp.hpp`

**Comment by WesselDeZeeuw** [comment]
Hi Joakim,

Thank you for your detailed feedback! I appreciate it. I will look at these comments for sure. I do have a few questions that will probably help me do these checks!

1.  What exactly is Jenkins build and how do I build this? 
2. I have based these codes largely on the code that was approved for the Carter-Tracy Aquifer. So I understand your comments and why I should make these enhancements, however I don't understand why some of these comments shouldn't be in the CT Aquifer code as well. For example the default handling. I don't understand why this must be done in this way when compared to the CT aquifer parsing where this isn't done.
3. I am actually unable to build the OPM simulator. I am missing parts of the Dune package (dune-uggrid. dune-fem and more) and it breaks. I installed dune from the OPM prerequisite page. Do you have a tip on how to overcome this problem? 
4. Is there a standard for the indentation size (because I think this is just a standard setting of my text handler) and the usage of the brackets? If so, what is it and how can I read about it?

Thanks for the answer, I will be busy with the changes. Also I will squash the commits into one as suggested by @alfbr !

**Comment by alfbr** [comment]
> I am actually unable to build the OPM simulator. I am missing parts of the Dune package (dune-uggrid. dune-fem and more) and it breaks. I installed dune from the OPM prerequisite page. Do you have a tip on how to overcome this problem?

The dune packages provided by OPM is sufficient for building opm-simulators, just remove any remnants of your self-compiled dune. In particular, you do not need dune-uggrid nor dune-fem.

You probably do need to download libecl from git though, as the packaged version will fail on recent git of opm-common. Just download it along side the opm-repositories, make sure all are up-to-date and to the cmake and make drill on all (start with libecl, then follow the dependency order for the opm-modules). Start with pristine repositories from git, and let us know of any problems.

Thanks for the credit, but I believe @atgeirr gave you the tip on squashing commits :-)

**Comment by joakim-hove** [comment]
> What exactly is Jenkins build and how do I build this?

Jenkins is a system for *continous integration*; jenkins will typically download software, build it and run all the tests. In the opm community we have set up a jenkins server here: https://ci.opm-project.org/ - that server communicates with github using a terse vocabulary like the: "jenkins build this please" a build test of your PR is started if you write a github comment with the exact content:
```txt
jenkins build this please
```
you will see that this PR failed - i.e. jenkins has come back with a :x:. You must be whitelisted te be able to initiate a jenkins build - that will hoepfully be handled by: @akva2.

**Comment by joakim-hove** [comment]
> I have based these codes largely on the code that was approved for the Carter-Tracy Aquifer. So I understand your comments and why I should make these enhancements, however I don't understand why some of these comments shouldn't be in the CT Aquifer code as well. 

You are absolutely right - the people writing the CT Aquifer code did not really listen to me. In all fairness it should be added that I misunderstood them at one point and gave some bad advice on pointer use.

**Comment by joakim-hove** [comment]
> Is there a standard for the indentation size (because I think this is just a standard setting of my text handler) and the usage of the brackets? If so, what is it and how can I read about it?

Braces and spaces and naming ... is a can of worms which people generally do not agree upon. But I think this is something everyone in OPM agrees more-or-less to:

1. Use four spaces for indentation.
2. Use spaces - not tabs.
3. Try to follow the surrounding code style for braces, naming and so on.

**Comment by joakim-hove** [comment]
> I am actually unable to build the OPM simulator. I am missing parts of the Dune package (dune-uggrid. dune-fem and more) and it breaks. I installed dune from the OPM prerequisite page. Do you have a tip on how to overcome this problem?

Please post more details here.

**Comment by WesselDeZeeuw** [comment]
> you will see that this PR failed - i.e. jenkins has come back with a ❌. You must be whitelisted te be able to initiate a jenkins build - that will hoepfully be handled by: @akva2.

So. how can you find out which tests have failed? I.e. does it also give information of which tests didn't pass so that you know where to improve your code?

**Comment by WesselDeZeeuw** [comment]
> You are absolutely right - the people writing the CT Aquifer code did not really listen to me. In all fairness it should be added that I misunderstood them at one point and gave some bad advice on pointer use.

Allright, so do you still want me to make these changes then? Or should I leave it?

**Comment by akva2** [comment]
hmm? click the details link, you have all the info you want and more.

**Comment by WesselDeZeeuw** [comment]
> Please post more details here.

Right. So I run the following build file created by @nairr. At first this worked, but now the build is interrupted. To be honest I don't really understand the errors, or how I should interpret it. I have added two pictures with the start and end of the erros. So I hope that this helps.

error1
error2
[build_allrepo.txt](https://github.com/OPM/opm-common/files/2433483/build_allrepo.txt)

**Comment by WesselDeZeeuw** [comment]
> hmm? click the details link, you have all the info you want and more.

Awesome! Thanks

**Comment by joakim-hove** [comment]
> Allright, so do you still want me to make these changes then? Or should I leave it?

Yes - you should make these changes.

**Comment by akva2** [comment]
we need more info than that. run the whole build process, using make VERBOSE=1, one build job and redirect everything to a log file, e.g.

```
make VERBOSE=1 2>&1 >> everything.log
```

my initial suspicion is that some of the repos are out of sync.

**Comment by WesselDeZeeuw** [comment]
> we need more info than that. run the whole build process, using make VERBOSE=1, one build job and redirect everything to a log file, e.g.
> make VERBOSE=1 2>&1 >> everything.log
>
> my initial suspicion is that some of the repos are out of sync.

Ok. Sorry to ask these perhaps dumb questions, but then the build file looks something like this right>

process.stdin.write('cd opm-common/' + ' \n')
process.stdin.write('cd ./build/' + ' \n')
process.stdin.write('make VERBOSE=1 2>&1+ ' \n') >> "everything.log"
I am not really sure how to do this. I have a lot to learn ;-)

**Comment by akva2** [comment]
you certainly cannot redirect stuff in python like that. why is that thing using python for shell interaction? seems rather counter-productive.

**Comment by WesselDeZeeuw** [comment]
> you certainly cannot redirect stuff in python like that. why is that thing using python for shell interaction? seems rather counter-productive.

I don't now.. @nairr explained that this shell would run the build files op the directories in the correct order. Then I build the entire thing by just running the python shell. I can also do it per directory by setting into the command window: 
"make -j 4" so I guess I have to change this to "make VERBOSE=1"?

**Comment by alfbr** [comment]
You seem to be building on Ubuntu. There really is no need for any bulid scripts there. install all dependencies (with the lone exception of libecl) from packages. Then you only need to be able to make a build folder, go into it, and run cmake and make. This is all explained here:
https://opm-project.org/?page_id=231
As long as all repositories are up to date, it works. I am happy to do a video session with you to explain how cmake and make works.

**Comment by akva2** [comment]
yes, add the VERBOSE=1, remove the -jx params and redirect output from the python script.

**Comment by WesselDeZeeuw** [comment]
> You seem to be building on Ubuntu. There really is no need for any bulid scripts there. install all dependencies (with the lone exception of libecl) from packages. Then you only need to be able to make a build folder, go into it, and run cmake and make. This is all explained here:
> https://opm-project.org/?page_id=231
> As long as all repositories are up to date, it works. I am happy to do a video session with you to explain how cmake and make works.

This would be very nice! :-) I guess some extra background is really helpful for me and for you as well! When do you have time? And how can I contact you?

**Comment by alfbr** [comment]
Hangouts is good, after lunch tomorrow.

@akva2 , you forgot to run cmake. Adding in a line for cmake, the for loop above builds all modules, and that is all you need.

**Comment by akva2** [comment]
i did not forget anything, the original code did not run cmake.

File: src/opm/parser/eclipse/EclipseState/Aquifetp.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  	{
+  		return m_aqufetp;
+  	}
+
+  	int Aquifetp::getAqPvtTabID(size_t aquiferIndex)
```
[the remark this answers is not in the document]
Comment (WesselDeZeeuw) [reply src/opm/parser/eclipse/EclipseState/Aquifetp.cpp]: What do you mean with this?

**Comment by alfbr** [comment]
> i did not forget anything, the original code did not run cmake.

Sorry, didn't look at the Python snippet. Cmake has obviously already been run from somewhere else.

See you tomorrow Wessel :-)

File: src/opm/parser/eclipse/EclipseState/Aquifetp.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Aquifetp.cpp]: This is just an opinion on my part; but the `Aquifetp` class is a container, the container should not really care about the innards of the objects it is holding. So I would say the best way to do this was:
```C++
Aquifetp aq_container( );
...
const auto& aq = container[index]; // requires you implement operator[] - which might be nice.
int aq_id = aq.aquiferID;
```
i.e. calling scope interacts with the `AquiferRecord` struct directly, not via the container. No big deal.

**Comment by WesselDeZeeuw** [comment]
In the category dare to ask: Is it possible to run Jenkins test myself locally on my PC? I dit find the opm-simulators/Jenkins/build.sh file but it said permission denied. Or should I push the changes, so that we can do it here (I actually haven't changed too much, just the basics for a successful compilation)

**Comment by bska** [comment]
> Is it possible to run Jenkins test myself locally on my PC?

Yes, you're supposed to be able to do that.  We have some information here,

https://wiki.opm-project.org/index.php?title=Jenkins_triggers#Using_the_scripts_locally

but I have not used that myself.  Our resident Jenkins expert is @akva2 so maybe he can fill in the details.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
What happened?

**Comment by WesselDeZeeuw** [comment]
I was getting a crazy error in my branch from the RestartIO function. I decided to start over and add the files one by one, to see where the error is. But then apparently I closed this... I am sorry, just starting to get the hang of things. Definately will not make this mistake again. I will do my adjustments then push it here, will this request then open again?

**Comment by joakim-hove** [comment]
>I am sorry, just starting to get the hang of things. Definately will not make this mistake again. I

No worries :-)

> will do my adjustments then push it here, will this request then open again?

Just push the "Reopen" button.

**Comment by WesselDeZeeuw** [comment]
Goodmorning everyone.
So yesterday in the evening I pushed another commit and I wanted to post this message already by then, but my computer didn't want to connect to the internet after this push so I interpreted this as a sign to go home ;-) 

But, together with @nairr we discussed the default handeling of the initial pressure p0. It was similar to the discussion in #476 by @kel85uk and I have decided to implement the idea of @joakim-hove as in #478. 

So this is another step further into the implementation of the fetkovich parsing. @joakim-hove because I made a slight error in closing this pull request, a lot of your comments also have vanished.. Could you give some more pointers on what's next?
Thanks!

P.S. I pushed this after I got a successful build and all the test passed, so if the Jenkins fails then please explain me how I can know this beforehand.

File: src/opm/parser/eclipse/EclipseState/Aquifetp.cpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            data.d0 = aqufetpRecord.getItem("DAT_DEPTH").getSIDouble(0);
+            data.C_t = aqufetpRecord.getItem("C_T").getSIDouble(0);
+            data.pvttableID = aqufetpRecord.getItem("TABLE_NUM_WATER_PRESS").template get<int>(0);
+	    data.V0 = aqufetpRecord.getItem("V0").getSIDouble(0);
+            data.J = aqufetpRecord.getItem("PI").getSIDouble(0);
+
+            if (aqufetpRecord.getItem("P0").hasValue(0) )
+	    {
+		double * raw_ptr = new double ( aqufetpRecord.getItem("P0").getSIDouble(0));
+		data.p0.reset( raw_ptr );
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Aquifetp.cpp]: Alternatively (I think) you can do:
```
data.p0 = std::make_shared<double>()
```

File: src/opm/parser/eclipse/EclipseState/Aquifetp.cpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            data.aquiferID = aqufetpRecord.getItem("AQUIFER_ID").template get<int>(0);
+            data.d0 = aqufetpRecord.getItem("DAT_DEPTH").getSIDouble(0);
+            data.C_t = aqufetpRecord.getItem("C_T").getSIDouble(0);
+            data.pvttableID = aqufetpRecord.getItem("TABLE_NUM_WATER_PRESS").template get<int>(0);
+	    data.V0 = aqufetpRecord.getItem("V0").getSIDouble(0);
+            data.J = aqufetpRecord.getItem("PI").getSIDouble(0);
+
+            if (aqufetpRecord.getItem("P0").hasValue(0) )
+	    {
+		double * raw_ptr = new double ( aqufetpRecord.getItem("P0").getSIDouble(0));
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Aquifetp.cpp]: Alternatively you can do:
```
data.p0 = std::make_shared<double>( aquferRecord.getItem("P0") ...)
```
in one statement.

**Comment by joakim-hove** [comment]
> Could you give some more pointers on what's next?

I think this is good; essentially ready for merge. Instead of returning the `std::vector<>` of records I would implement `begin()` and `end()`iterators (which of course just forward to the underlying `std::vector<>` - but that is detail. 

Still some weird indentations here and there; but has never killed anyone.

> I pushed this after I got a successful build and all the test passed, so if the Jenkins fails then please explain me how I can know this beforehand.

Well - the ways of the Jenkins lord are mysterious; the most typical reason it might work at your place and not with Jenkins is that the some modules have been updated and your local copy is not 100% upto date. Furthermore Jenkins also runs all the MPI tests which many people do not do locally. But all in all - if you are all green locally that it is at least very close to all green on Jenkins.

I'll fire Jenkins now; if that is all green I am happy to merge this.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
> Should I overload the == operator in the class file?

Look like you are trying to compare `std::shared_ptr<double>` and `false`? I do not think you can do that.

**Comment by WesselDeZeeuw** [comment]
> jenkins does nothing special. either you are not building that file on your machine, or there are compiler and/or boost version differences.

I see now that Jenkins is using boost version 1.62.0, and I am running version 1.54.0 how do I update this?

> Maybe you can compare it with nullptr.

Thanks! Good tip!

File: tests/parser/AquifetpTests.cpp [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    std::vector< Aquifetp::AQUFETP_data > aquifetp_null = init_aquifetp(aqufetp_deck_null);
+    BOOST_CHECK_EQUAL(aquifetp_null.size(), 0);    
+
+    auto aqufetp_deck_default = createAquifetpDeck_defaultPressure();
+    std::vector< Aquifetp::AQUFETP_data > aquifetp_default = init_aquifetp(aqufetp_deck_default);
+    for (const auto& it : aquifetp_default){
+        BOOST_CHECK_EQUAL(it.aquiferID , 1);
+	BOOST_CHECK_EQUAL(it.V0, 2.0e9);
+	BOOST_CHECK_EQUAL(it.J, 500/86400e5);  
+	BOOST_CHECK(it.p0 == false);
```
Comment (joakim-hove) [inline tests/parser/AquifetpTests.cpp]: What you want here is something which *evaluates* to `false` - which is not the same as `== false`. I guess this should do the trick:
```
BOOST_CHECK( !it.p0 );
```

Comment (WesselDeZeeuw) [reply tests/parser/AquifetpTests.cpp]: I changed this particular line :-)

**Comment by WesselDeZeeuw** [comment]
Could someone fire jenkins for me. I changed the lines where jenkins failed so I am curious if it breaks somewhere else now! Thanks

**Comment by WesselDeZeeuw** [comment]
Could someone fire jenkins for me. I changed the lines where jenkins failed so I am curious if it breaks somewhere else now! Thanks

**Comment by WesselDeZeeuw** [comment]
Could someone please fire jenkins for me? I changed the lines which made it break and I am curious to see if there are more places where jenkins fails. Locally everything builds and all the tests pass. Thanks

**Comment by WesselDeZeeuw** [comment]
Could someone please fire jenkins for me? I changed the lines which made it break and I am curious to see if there are more places where jenkins fails. Locally everything builds and all the tests pass. Thanks

**Comment by WesselDeZeeuw** [comment]
Could someone please fire jenkins for me? I changed the lines which made it break and I am curious to see if there are more places where jenkins fails. Locally everything builds and all the tests pass. Thanks

**Comment by WesselDeZeeuw** [comment]
Could someone please fire jenkins for me? I changed all the lines which made it break but I am curious to see if there are more errors there. Locally everything builds and the tests pass 100%. Thanks

**Comment by WesselDeZeeuw** [comment]
Could someone please fire jenkins for me? I changed all the lines which made it break but I am curious to see if there are more errors there. Locally everything builds and the tests pass 100%. Thanks

**Comment by WesselDeZeeuw** [comment]
Could someone please fire jenkins for me? I changed all the lines which made it break but I am curious to see if there are more errors there. Locally everything builds and the tests pass 100%. Thanks

**Comment by bska** [comment]
jenkins build this please

**Comment by andlaus** [comment]
jenkins build this please

**Comment by andlaus** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by andlaus** [comment]
jenkins build this please

**Comment by andlaus** [comment]
sorry for the noise, github seems to have some serious problems with its comments system at the moment.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by WesselDeZeeuw** [comment]
Awesome! So it looks like it passed. Is there anything from my side I need to do right now? I still have two small remarks on making the style a bit more appropriate, but I guess I can (re)open the pull request right?

**Comment by joakim-hove** [comment]
> Is there anything from my side I need to do right now?

This *can* be merged as is; I would prefer if you removed the `Aquifetp::getAquifers()` method and instead added `begin()` and `end()` methods (which just forward directly to underlying `std::vector`); but it is not essential.

> but I guess I can (re)open the pull request right?

You can always make *a new* pull request; it is not customary to reopen a PR with additional changes after it has been merged - although I assume it is technically possible to do it that way.

**Comment by joakim-hove** [comment]
Congratulation on your first accepted PR.

**Comment by WesselDeZeeuw** [comment]
Thank you very much! On to the next one !
